File: mcp-wrapper/scripts/import-audit.mjs
#!/usr/bin/env node
// Static guard: scan the built `dist/` tree for forbidden platform-specific
// module imports. Spec §9.3 requires "Wrapper has zero platform-specific
// imports" — this script enforces that at build verification time, runnable
// outside vitest so CI / hand-checks can invoke it directly.
//
// Usage:
//   node scripts/import-audit.mjs [dist-dir]
//
// Default dist-dir is `<package-root>/dist`. The companion vitest test
// `tests/no-platform-imports.test.ts` (#58) does a substring-level grep over
// the whole compiled file; this script narrows that to the actual import
// path so a tool description string mentioning e.g. "Discord" doesn't trip
// the audit.
//
// Exit codes:
//   0 — clean
//   1 — one or more forbidden imports found (printed to stderr)
//   2 — usage / IO error (e.g. dist dir missing)

import { existsSync, readFileSync, readdirSync, statSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const FORBIDDEN_TOKENS = ['discord', 'applescript', 'osascript', 'chat.db', 'imessage'];

// Regexes that capture the module path (group 1) of every JS import form
// `tsc` can emit. We intentionally match only on the *module specifier* and
// never on arbitrary string literals.
//
// Covered forms:
//   import 'mod';
//   import x from 'mod';
//   import { a, b } from 'mod';
//   import * as ns from 'mod';
//   import x, { y } from 'mod';
//   export { a } from 'mod';
//   export * from 'mod';
//   export * as ns from 'mod';
//   await import('mod');                        — dynamic, statically analyzable
//   require('mod');                              — interop / CJS shim emitted by tsc
const IMPORT_PATTERNS = [
  /\bimport\s+(?:[\w*${}\s,]+\s+from\s+)?['"]([^'"]+)['"]/g,
  /\bexport\s+(?:\*(?:\s+as\s+\w+)?|\{[^}]*\})\s+from\s+['"]([^'"]+)['"]/g,
  /\bimport\s*\(\s*['"]([^'"]+)['"]\s*\)/g,
  /\brequire\s*\(\s*['"]([^'"]+)['"]\s*\)/g,
];

function walkJs(dir) {
  const out = [];
  for (const name of readdirSync(dir)) {
    const full = join(dir, name);
    const s = statSync(full);
    if (s.isDirectory()) {
      out.push(...walkJs(full));
    } else if (name.endsWith('.js')) {
      out.push(full);
    }
  }
  return out;
}

/**
 * Extract import module specifiers from a JS source string. Returns each
 * specifier paired with the 1-based line it appeared on (best-effort, used
 * only for human-readable error output).
 */
export function extractImportSpecifiers(source) {
  const hits = [];
  for (const pattern of IMPORT_PATTERNS) {
    // Reset lastIndex defensively — these are module-level regexes with /g.
    pattern.lastIndex = 0;
    let match;
    while ((match = pattern.exec(source)) !== null) {
      const specifier = match[1];
      const line = source.slice(0, match.index).split('\n').length;
      hits.push({ specifier, line });
    }
  }
  return hits;
}

/**
 * Check a single specifier against the forbidden-token list. Returns the
 * matched token, or `null` if clean. Comparison is case-insensitive to match
 * the no-platform-imports.test.ts behavior.
 */
export function findForbiddenToken(specifier, tokens = FORBIDDEN_TOKENS) {
  const lower = specifier.toLowerCase();
  for (const token of tokens) {
    if (lower.includes(token)) {
      return token;
    }
  }
  return null;
}

/**
 * Audit a directory of compiled `.js` files. Returns the list of violations.
 * Each violation: { file, specifier, token, line }.
 */
export function auditDist(distDir) {
  const violations = [];
  const files = walkJs(distDir);
  for (const file of files) {
    const source = readFileSync(file, 'utf8');
    for (const { specifier, line } of extractImportSpecifiers(source)) {
      const token = findForbiddenToken(specifier);
      if (token !== null) {
        violations.push({ file, specifier, token, line });
      }
    }
  }
  return violations;
}

function defaultDistDir() {
  const here = dirname(fileURLToPath(import.meta.url));
  // scripts/ sits under the package root next to dist/.
  return resolve(here, '..', 'dist');
}

function main(argv) {
  const explicit = argv[2];
  const distDir = explicit ? resolve(explicit) : defaultDistDir();

  if (!existsSync(distDir)) {
    process.stderr.write(
      `[import-audit] dist directory not found: ${distDir}\n` +
        `Run \`npm run build\` first, or pass a path: ` +
        `node scripts/import-audit.mjs <dist-dir>\n`,
    );
    return 2;
  }

  const stat = statSync(distDir);
  if (!stat.isDirectory()) {
    process.stderr.write(`[import-audit] not a directory: ${distDir}\n`);
    return 2;
  }

  const violations = auditDist(distDir);
  if (violations.length === 0) {
    process.stdout.write(`[import-audit] OK — no forbidden imports under ${distDir}\n`);
    return 0;
  }

  process.stderr.write(
    `[import-audit] FAIL — ${violations.length} forbidden import(s) found:\n`,
  );
  for (const v of violations) {
    process.stderr.write(
      `  ${v.file}:${v.line}  imports "${v.specifier}"  (matched token: "${v.token}")\n`,
    );
  }
  return 1;
}

// Run only when invoked as a script, not when imported by tests.
const invokedAsScript = (() => {
  if (typeof process === 'undefined' || !process.argv[1]) return false;
  try {
    return import.meta.url === new URL(`file://${process.argv[1]}`).href;
  } catch {
    return false;
  }
})();

if (invokedAsScript) {
  process.exit(main(process.argv));
}

File: mcp/scripts/import_audit.py
#!/usr/bin/env python3
"""Static guard: scan the ``amc_mcp`` source tree for forbidden imports.

Spec §9.3 requires "Wrapper has zero platform-specific imports" — this
script is the static enforcement, runnable outside pytest so CI / hand
checks can invoke it directly.

Usage::

    python scripts/import_audit.py [package-dir]

Default package dir is ``<repo>/src/amc_mcp``. The audit parses each ``.py``
with ``ast`` and inspects only module specifiers — a docstring or tool
description that mentions e.g. "Discord" never produces a false positive.

Exit codes::

    0 — clean
    1 — one or more forbidden imports found (printed to stderr)
    2 — usage / IO error (e.g. package dir missing)
"""

from __future__ import annotations

import ast
import sys
from collections.abc import Iterable
from dataclasses import dataclass
from pathlib import Path

FORBIDDEN_TOKENS: tuple[str, ...] = (
    "discord",
    "applescript",
    "osascript",
    "chat.db",
    "imessage",
)


@dataclass(frozen=True)
class Violation:
    file: Path
    specifier: str
    token: str
    line: int


def extract_import_specifiers(source: str) -> list[tuple[str, int]]:
    """Return ``(specifier, lineno)`` for every ``import`` / ``from … import``."""
    try:
        tree = ast.parse(source)
    except SyntaxError:
        return []
    out: list[tuple[str, int]] = []
    for node in ast.walk(tree):
        if isinstance(node, ast.Import):
            for alias in node.names:
                out.append((alias.name, node.lineno))
        elif isinstance(node, ast.ImportFrom) and node.module:
            out.append((node.module, node.lineno))
    return out


def find_forbidden_token(specifier: str, tokens: Iterable[str] = FORBIDDEN_TOKENS) -> str | None:
    """Return the first matching forbidden token in ``specifier``, or None."""
    lower = specifier.lower()
    for token in tokens:
        if token in lower:
            return token
    return None


def audit_package(package_dir: Path) -> list[Violation]:
    violations: list[Violation] = []
    for path in sorted(package_dir.rglob("*.py")):
        try:
            source = path.read_text(encoding="utf-8")
        except OSError:
            continue
        for specifier, line in extract_import_specifiers(source):
            token = find_forbidden_token(specifier)
            if token is not None:
                violations.append(Violation(file=path, specifier=specifier, token=token, line=line))
    return violations


def _default_package_dir() -> Path:
    here = Path(__file__).resolve().parent
    return here.parent / "src" / "amc_mcp"


def main(argv: list[str]) -> int:
    explicit = argv[1] if len(argv) > 1 else None
    package_dir = Path(explicit).resolve() if explicit else _default_package_dir()

    if not package_dir.exists():
        sys.stderr.write(
            f"[import-audit] package directory not found: {package_dir}\n"
            "Pass an explicit path: python scripts/import_audit.py <package-dir>\n"
        )
        return 2
    if not package_dir.is_dir():
        sys.stderr.write(f"[import-audit] not a directory: {package_dir}\n")
        return 2

    violations = audit_package(package_dir)
    if not violations:
        sys.stdout.write(f"[import-audit] OK — no forbidden imports under {package_dir}\n")
        return 0

    sys.stderr.write(f"[import-audit] FAIL — {len(violations)} forbidden import(s) found:\n")
    for v in violations:
        sys.stderr.write(
            f"  {v.file}:{v.line}  imports {v.specifier!r}  (matched token: {v.token!r})\n"
        )
    return 1


if __name__ == "__main__":
    sys.exit(main(sys.argv))
